src/github.com/getlantern/whitelist/whitelist.go

```diff
@@ -0,0 +1,239 @@
+		}
+	}
+	return entries
+}
+
+func (wl *Whitelist) addOriginal() []string {
+	wl.entries = LoadDefaultList()
+	return wl.entries
+}
+
+func (wl *Whitelist) Copy() *Config {
+	return &Config{
+		Additions: wl.cfg.Additions,
+		Deletions: wl.cfg.Deletions,
+		Cloud:     wl.cfg.Cloud,
+	}
+}
+
+func (wl *Whitelist) GetConfig() *Config {
+	return wl.cfg
+}
+
+/* This function calculaties the delta additions and deletions
+ * to the global whitelist; these changes are then propagated
+ * to the PAC file
+ */
+func (wl *Whitelist) UpdateEntries(entries []string) []string {
+	log.Debug("Updating whitelist entries...")
+
+	toAdd := set.New()
+
+	for i := range entries {
+		toAdd.Add(entries[i])
+	}
+
+	/* whitelist customizations */
+	toRemove := set.Difference(wl.cloudSet, toAdd)
+	wl.cfg.Deletions = set.StringSlice(toRemove)
+
+	/* new entries are any new domains the user wishes
+	* to proxy that weren't found on the global whitelist
+	* already
+	 */
+	newEntries := set.Difference(toAdd, wl.cloudSet)
+	wl.cfg.Additions = set.StringSlice(newEntries)
+	wl.entries = set.StringSlice(toAdd)
+	go wl.updatePacFile()
+
+	return wl.entries
+}
+
+func (wl *Whitelist) updatePacFile() (err error) {
+
+	pacFile := &PacFile{}
+
+	pacFile.file, err = os.Create(PacFilePath)
+	defer pacFile.file.Close()
+	if err != nil {
+		log.Errorf("Could not create PAC file")
+		return
+	}
+	/* parse the PAC file template */
+	pacFile.template, err = template.ParseFiles(PacTmpl)
+	if err != nil {
+		log.Errorf("Could not open PAC file template: %s", err)
+		return
+	}
+
+	log.Debugf("Updating PAC file; path is %s", PacFilePath)
+	pacFile.l.Lock()
+	defer pacFile.l.Unlock()
+
+	data := make(map[string]interface{}, 0)
+	data["Entries"] = wl.entries
+	err = pacFile.template.Execute(pacFile.file, data)
+	if err != nil {
+		log.Errorf("Error generating updated PAC file: %s", err)
+	}
+
+	return err
+}
+
+func (wl *Whitelist) GetEntries() []string {
+	return wl.entries
+}
+
+func ParsePacFile() *Whitelist {
+	wl := &Whitelist{}
+
+	log.Debugf("PAC file found %s; loading entries..", PacFilePath)
+	/* pac file already present */
+	program, err := parser.ParseFile(nil, PacFilePath, nil, 0)
+	/* otto is a native JavaScript parser;
+	we just quickly parse the proxy domains
+	from the PAC file to
+	cleanly send in a JSON response
+	*/
+	vm := otto.New()
+	_, err = vm.Run(program)
+	if err != nil {
+		log.Errorf("Could not parse PAC file %+v", err)
+		return nil
+	} else {
+		value, _ := vm.Get("proxyDomains")
+		log.Debugf("PAC entries %+v", value.String())
+		if value.String() == "" {
+			/* no pac entries; return empty array */
+			wl.entries = []string{}
+			return wl
+		}
+
+		/* need to remove escapes
+		* and convert the otto value into a string array
+		 */
+		re := regexp.MustCompile("(\\\\.)")
+		list := re.ReplaceAllString(value.String(), ".")
+		wl.entries = strings.Split(list, ",")
+		log.Debugf("List of proxied sites... %+v", wl.entries)
+	}
+	return wl
+}
```

Comment:
This is mainly used to test the contents of the PAC file and could probably be moved to whitelist_test.go